File: field_force/field_force/page/customer_wise_delivery/customer_wise_delivery.js
frappe.pages['customer-wise-delivery'].on_page_load = function(wrapper) {
	var page = frappe.ui.make_app_page({
		parent: wrapper,
		title: 'Customer Wise Delivery Report',
		single_column: true
	});
	set_background_color();
	new CustomerWiseDelivery(page);
  }

  class CustomerWiseDelivery {
	constructor(page) {
	  this.page = page;
	  this.make_form();
	  this.make_menu();
	}

	initialize_events = () => {
	  	this.fetch_and_render();
	}

	make_menu = () => {
		this.page.add_action_icon("refresh", () => {
			this.fetch_and_render(true);
		}, "Refresh");

    if (frappe.user.has_role("Warehouse User") || frappe.user.has_role("System Manager")) {
		this.page.add_button("Update Status", () => {
			let trips = [];
			let checkboxes = document.querySelectorAll('input[type="checkbox"]');
			checkboxes.forEach(function (checkbox) {
				if (checkbox.checked) {
					trips.push(checkbox.id);
				}
			});
			if (trips.length > 0) {
				let d = new frappe.ui.Dialog({
					title: 'Enter details',
					fields: [
						{
							fieldtype: "Select",
							label: __("Status"),
							fieldname: "status",
							options: [
								"",
								"Draft",
								"Scheduled",
								"In Transit",
								"Completed",
								"Cancelled"
							],
						},
					],
					size: 'small', // small, large, extra-large 
					primary_action_label: 'Update',
					primary_action(values) {
						let trips = [];
						let checkboxes = document.querySelectorAll('input[type="checkbox"]');
						checkboxes.forEach(function (checkbox) {
							if (checkbox.checked) {
								trips.push(checkbox.id);
							}
						});
						frappe.call({
							method:'field_force.field_force.page.customer_wise_delivery.customer_wise_delivery.update_delivery_stop_status',
							args: {
								trips: trips,
								value:values
							},
							callback: (r) => {
							}
						});
						d.hide();
					}
				});
				
				d.show();
			}else{

				msgprint("Select Delivery Stops First","Message")
			}
				
			})
		}
	}
	make_form = () => {
	  	this.form = new frappe.ui.FieldGroup({
			fields: [
				{

					fieldname: 'from_date',
					label: __('From Date'),
					fieldtype: 'Date',
					default: frappe.datetime.add_days(frappe.datetime.get_today(), -30),
					change: () => this.fetch_and_render()
				},
				{
					fieldtype: 'Column Break'
				},
				{
					fieldname: 'to_date',
					label: __('To Date'),
					fieldtype: 'Date',
					default: frappe.datetime.get_today(),
					change: () => this.fetch_and_render(),
				},
				{
					fieldtype: 'Column Break'
				},
				{
					fieldname:"customer",
					label: __("Customer"),
					fieldtype: "Link",
					options: "Customer",
					change: () => this.fetch_and_render(),
				},
				{
					fieldtype: 'Column Break'
				},
				{
					fieldname: 'delivery_trip',
					label: __('ID'),
					fieldtype: 'Link',
					options: "Delivery Trip",
					change: () => this.fetch_and_render(),
				},
				{
					fieldtype: 'Column Break'
				},
				{

					fieldname: 'territory',
					label: __('Territory'),
					fieldtype: 'Link',
					options: "Territory",
					change: () => this.fetch_and_render()
				},
				{
					fieldtype: 'Column Break'
				},
				{

					fieldname: 'status',
					label: __('Status'),
					fieldtype: 'Link',
					options: "Delivery Trip Status",
					change: () => this.fetch_and_render()
				},
				{
					fieldtype: 'Section Break'
				},
				{
					fieldname:"vehicle",
					label: __("Vehice"),
					fieldtype: "Link",
					options: "Vehicle",
					change: () => this.fetch_and_render(),
				},
				{
					fieldtype: 'Column Break'
				},
				{
					fieldname:"driver",
					label: __("Driver"),
					fieldtype: "Link",
					options: "Driver",
					change: () => this.fetch_and_render(),
				},
				{
					fieldtype: 'Column Break'
				},
				{
					fieldname: 'scrollable_view',
					label: __('Scrollable View'),
					fieldtype: 'Check',
					default: 0,
					change: () => this.fetch_and_render(),
					hidden: false
				},
				{
					fieldtype: 'Column Break'
				},
				{
					fieldtype: 'Column Break'
				},
				{
					fieldtype: 'Section Break'
				},
				{
					fieldtype: 'HTML',
					fieldname: 'preview'
				}
			],
			body: this.page.body
		});
		this.form.make();
		this.fetch_and_render()
	}
	fetch_and_render = (refresh=false) => {
    let filters= this.form.get_values();
    let queryParams = window.location.search;

    if (!filters.status){
		if (queryParams) {
      const urlParams = new URLSearchParams(queryParams);
      console.log(urlParams);
      console.log(urlParams.has('workflow_state'));
      console.log(urlParams.has('delivery_trip_created'));

      if (urlParams.has('workflow_state')){
        let workflowStateValue = urlParams.get('workflow_state');
        workflowStateValue = decodeURIComponent(workflowStateValue);
        filters['status'] = workflowStateValue
        $('[data-fieldname="status"]').val(workflowStateValue)
      }

      if (urlParams.has('delivery_trip_created')){
        let deliveryTripCreated = urlParams.get('delivery_trip_created');
        deliveryTripCreated = decodeURIComponent(deliveryTripCreated);
        filters['delivery_trip_created'] = parseInt(deliveryTripCreated)
        $('[data-fieldname="delivery_trip_created"]').val(deliveryTripCreated)
      }

			window.history.replaceState({}, document.title, window.location.pathname);
		}
  }



	  // to prevent the duplicate call
	  if (refresh === false && JSON.stringify(this.filters) === JSON.stringify(filters)){
		return;
	  }
	  this.filters = this.form.get_values();

	  if (!filters.from_date) {
		this.form.get_field('preview').html('');
		return;
	  }

	  this.form.get_field('preview').html(`
			  <div class="text-muted margin-top">
				  ${__("Fetching...")}
			  </div>
		  `);
	  frappe.call('field_force.field_force.page.customer_wise_delivery.customer_wise_delivery.execute', {
		filters: filters,
		freeze: true
	  }).then(r => {
		let diff = r.message[0];
		let fields = r.message[1];
		this.render(diff, fields);
	  });
	}

	render = (diff, fields) => {
	  let table_header = this.table_header(fields);
	  let table_body = this.table_body(diff, fields);
	  let scrollable_view = this.form.get_values()['scrollable_view']

	  if (scrollable_view === 1) {
		this.form.get_field('preview').html(
		  `<div style="width:100%; overflow-x:auto; margin-top: -15px;">
			  <table class="table table-bordered" id="export_excel" style="width:160%">
					${table_header}${table_body}
			  </table>
			</div>
		  `);
	  } else {
		this.form.get_field('preview').html(
		`<div>
			  <table class="table table-bordered" id="export_excel">
					${table_header}${table_body}
			  </table>
			</div>
		`);
	  }

	}
	table_header = (headers) => {
	  let table_header = `<thead><tr>`;

	  headers.forEach(function (field, index) {
		table_header += `<th scope="col" style="width:${field.width}px !important;">${field.label || ''}</th>`;
	  })

	  table_header += `</tr>`;
	  return table_header;
	}

	table_body = (diff, fields) => {
	  var html = `<tbody>`;
	  var serial_number = 1;

	  diff.forEach(function (data, index) {
		html += `<tr>`;

		fields.forEach(function (field, indx) {
		  if (field.fieldname === 'sl') {
			data['sl'] = index + 1;
			html += get_absolute_format_and_html(field, data);
		  }
		else if (field.editable) {
				html += `<td id="${data.docname}_${field.fieldname}" class="editable-field parent-container"
						data-fieldname="${field.fieldname}" data-name="${data.docname}">${data[field.fieldname] || ''}</td>`
			}
		else if (field.fieldname === 'check') {
			if (frappe.user.has_role("Customer") && !frappe.user.has_role("System Manager")) {
				html += `<td style="padding-right: 4px;padding-left: 5px;"></td>`
			}else{
				html += `<td style="padding-right: 4px;padding-left: 5px;">
						<input type="checkbox" id="${data.docname}&${data.ds_docname}" 
						style="margin-right: 0px !important;margin-top: 12px;"></td>`
			}
			}
		else if (field.fieldname === 'delivery_trip_created') {
			let checked = data[field.fieldname] === 1 ? "checked" : "";
				html += `<td style="padding-right: 4px;padding-left: 5px;">
					<input type="checkbox" ${checked} id="${data.docname}" onclick="return false;"
					style="margin-right: 0px !important; accent-color: #00b2ff !important;"></td>`
		  }
		  else {
			html += get_absolute_format_and_html(field, data);
		  }
		})

		html += `</tr>`;
		serial_number = serial_number + 1;
	  })

	  html += `</tbody>`;
	  return html
	}
  }

//   amjad kalir bazar life general hospital
  function play_action(action) {
	frappe.xcall("field_force.field_force.page.customer_wise_delivery.customer_wise_delivery.play_action",
		{action: action}
	  ).then((response) => {
		let statusFieldID = response.docname + '_status';
		document.getElementById(response.docname).innerHTML = response.action;
		document.getElementById(statusFieldID).innerHTML = response.status;
	  });
  }

  function get_absolute_format_and_html(field, data, width = 50) {
	let docname = data.docname;
	let value = data[field.fieldname];

	if (field.fieldtype === "Image") {
	  return get_image_html(value);
	} else if (field.fieldtype === "Currency") {
	  return get_currency_format(value);
	} else if (field.fieldtype === "Data") {
	  return `<td id="${docname}_${field.fieldname}">${value || ''}</td>`;
	} else {
	  return `<td id="${docname}_${field.fieldname}">${value || ''}</td>`;
	}
  }

  function get_image_html(image_url) {
	return `
		  <td style="height:100px; width:120px;">
			  <a href="#">
				  <img style="height:100%; width:100%" id="image" src="${image_url}" onclick="(
					  function(e){
						  document.getElementById(\'modal_section\').style.display=\'block\';
						  var nAgt = navigator.userAgent;
						  if (nAgt.indexOf('Safari') !== -1) {
							  document.getElementById(\'img\').src=e.target.currentSrc;
						  }else{
							  document.getElementById(\'img\').src=e.path[0].currentSrc;
						  }
						  return false;
					  }
				  )(arguments[0]);
				  return false;
			  </a>
		  </td>`;
  }

  function get_currency_format(value) {
	const company = frappe.defaults.get_user_default("company");
	const currency = frappe.get_doc(":Company", company).default_currency;
	let value_in_currency = format_currency(value, currency);
	return `<td>${value_in_currency || ''}</td>`
  }

  function set_background_color(){
	let current_theme = document.documentElement.getAttribute("data-theme")
	if (current_theme === 'light'){
		$('.page-body').css('background', '#FFFFFF');
	} else{
		$('.page-body').css('background', 'black');
	}
  }
  function dateFormat(input_D, format_D) {
    const date = new Date(input_D);
    const day = date.getDate();
    const month = date.getMonth() + 1;
    const year = date.getFullYear();
    format_D = format_D.replace("MM", month.toString().padStart(2,"0"));
    if (format_D.indexOf("yyyy") > -1) {
        format_D = format_D.replace("yyyy", year.toString());
    } else if (format_D.indexOf("yy") > -1) {
        format_D = format_D.replace("yy", year.toString().substr(2,2));
    }
    format_D = format_D.replace("dd", day.toString().padStart(2,"0"));
    return format_D;
}

File: field_force/field_force/page/customer_wise_delivery/customer_wise_delivery.py
import frappe
import json
import datetime
from frappe import _
from frappe.utils import pretty_date
from field_force.field_force.page.utils import *
from field_force.field_force.page.utils import generate_excel_and_download, get_time_in_12_hour_format

@frappe.whitelist()
def execute(filters=None):
    columns = get_columns()
    data = get_absolute_data(filters)
    return data, columns

def get_absolute_data(filters, export=False):
    query_result = get_query_data(filters)

    for requisition in query_result:
        
        requisition['status'] = get_workflow_state_with_color(requisition)
        requisition['docname'] = requisition['id']
        requisition['ds_docname'] = requisition['ds_id']
        requisition['departure_date'] = frappe.format(requisition.departure_date, 'Date')
        change_color(requisition,"departure_date","#9b00d7")
        change_color(requisition,"contact_person","#00bf46")
        change_color(requisition,"contact_number","#8000cc")
        change_color(requisition,"customer_primary_address","#0600cc")

        if not export:
            user_role = frappe.get_roles(frappe.session.user)
            if "Customer" in user_role and "System Manager" not in user_role:
                change_color(requisition,"id","blue")
            else:
                set_link_to_doc(requisition, 'id','delivery-trip',val=requisition.id, color='blue')
                
            set_link_to_doc(requisition, 'requisition','requisition',val=requisition.requisition, color='#00918a')
            set_link_to_doc(requisition, 'customer','customer',val=requisition.customer, color='#0080ff')
            set_link_to_doc(requisition, 'territory','territory',val=requisition.territory, color='#0080ff	')
            set_link_to_doc(requisition, 'driver_name','driver',val=requisition.driver, color='#0080ff	')
            if requisition.get("vehicle"):
                vehicle = frappe.get_doc("Vehicle", requisition.get("vehicle"))
                labels = vehicle.name +","+ vehicle.model
                set_link_to_doc(requisition, 'vehicle','vehicle',val=vehicle.name,label=labels,color='#00918a')

        else:
            requisition['status'] = requisition.status

    return query_result

def get_conditions(filters):
    if not filters: filters = {}
    conditions = ''

    if filters.get("from_date"):
        conditions += ' and date(dt.departure_time) >= "{}"'.format(filters.get("from_date"))
    if filters.get("to_date"):
        conditions += ' and date(dt.departure_time) <= "{}"'.format(filters.get("to_date"))
    if filters.get("customer"):
        conditions += ' and ds.customer = "{}"'.format(filters.get("customer"))
    if filters.get("territory"):
        conditions += ' and dt.territory = "{}"'.format(filters.get("territory"))
    if filters.get("driver"):
        conditions += ' and dt.driver = "{}"'.format(filters.get("driver"))
    if filters.get("status"):
        conditions += ' and ds.status = "{}"'.format(filters.get("status"))
    if filters.get("vehicle"):
        conditions += ' and dt.vehicle = "{}"'.format(filters.get("vehicle"))
    if filters.get("requisition"):
        conditions += ' and ds.requisition = "{}"'.format(filters.get("requisition"))
    
    user_roles = frappe.get_roles(frappe.session.user)
    if "Customer" in user_roles and "System Manager" not in user_roles:
        conditions += ' and req.owner = "{}"'.format(frappe.session.user)
    return conditions


def get_query_data(filters):
    try:
        filters = json.loads(filters)
    except:
        pass

    conditions = get_conditions(filters)


    customers = frappe.get_list("Customer", filters={"disabled": 0}, pluck='name')
    customers_in_tuple_str = '("' + '", "'.join(customers) + '")'

    query = frappe.db.sql('''select dt.name as id, dt.driver_name, dt.driver, dt.territory, ds.requisition,
                        date(dt.departure_time) as departure_date, dt.vehicle, time(dt.departure_time) as departure_time,ds.requisition,
                         ds.status, ds.customer, ds.contact_person,ds.contact_number, ds.customer_primary_address, ds.total_qty,ds.name as ds_id
                        from `tabDelivery Stop` as ds left join `tabDelivery Trip` as dt on dt.name = ds.parent
                        left join `tabRequisition` as req on ds.requisition = req.name
                        where dt.name is not null and ds.customer in {} {}
                        order by departure_time desc'''.format(customers_in_tuple_str, conditions),as_dict=1)
    
    return query

def get_columns():
    columns =  [
        {'fieldname': 'check', 'label': '',"fieldtype":"Check", 'expwidth': 5, 'width': 20},
        {"label": _("ID"), "fieldtype": "Data", "fieldname": "id", "options": "Customer", "width": 120},
        {"label": _("Requisition"), "fieldtype": "Data", "fieldname": "requisition", "width": 120},
        {"label": _("Customer"), "fieldtype": "Data", "fieldname": "customer", "options": "Customer", "width": 180},
        {"label": _("Dep. Date"),"fieldtype": "Data","fieldname": "departure_date","width": 100},
        {"label": _("Territory"),"fieldtype": "Data","fieldname": "territory","options":"Territory","width": 120},
        {"label": _("Status"), "fieldtype": "Data","fieldname": "status", "width": 80},
        {"label": _("Vehicle"),"fieldtype": "Data","fieldname": "vehicle","width": 120},
        {"label": _("Driver Name"), "fieldtype": "Data", "fieldname": "driver_name", "width": 120},
        {"label": _("Address"),"fieldtype": "Data","fieldname":"customer_primary_address","width": 200},
        {"label": _("Contact Person"), "fieldtype": "Data", "fieldname": "contact_person", "width": 80},
        {"label": _("Contact Number"), "fieldtype": "Data", "fieldname": "contact_number", "width": 80},
        {"label": _("QTY"), "fieldtype": "Float", "fieldname": "total_qty", "width": 80},
    ]
    return columns
def get_appropiate_action_button(requisition):
    action = f'''<a href="/app/requisition/{requisition.get("name")} "id="{requisition.get("name")}_Approve" target="_blank" class="btn btn-success btn-sm"
                            style="width:50px;">View</a><br>'''
    return action

@frappe.whitelist()
def play_action(action):
    if action:
        requisition_name, action = tuple(action.split('_'))
        requisition = frappe.get_doc("Requisition", requisition_name)

        if requisition.workflow_state == 'Pending for Ops Team':
            if action == "Approve":
                requisition.workflow_state = 'Pending for Customer'
            else:
                requisition.workflow_state = 'Rejected by Ops Team'

        elif requisition.workflow_state == 'Pending for Customer':
            if action == "Approve":
                requisition.workflow_state = 'Approved'
            else:
                requisition.workflow_state = 'Rejected by Customer'


        elif requisition.workflow_state == 'Approved' and \
            frappe.has_permission("Requisition", "cancel", requisition.name) and action == 'Cancel':
            frappe.db.set_value("Requisition", requisition.name, 'workflow_state', 'Cancelled')
            frappe.db.set_value("Requisition", requisition.name, 'status', 'Cancelled')
            frappe.db.commit()

            requisition = frappe.get_doc("Requisition", requisition_name)
            requisition.cancel()

            return {
                'docname': requisition_name,
                'status': get_workflow_state_with_color(requisition),
                'action': ''
            }

        requisition.save()

        if requisition.workflow_state == "Approver" and requisition.docstatus == 0:
            requisition.submit()

        return {
            'docname': requisition_name,
            'status': get_workflow_state_with_color(requisition),
            'action': get_appropiate_action_button(requisition)
        }

def get_workflow_state_with_color(requisition):
    color = {
        "Scheduled" : "#ff5407",
        "In Transit" : "#005cef",
        "Completed" : "green",
        "Cancelled" : "red",
        "Draft" : "red"
    }
    return f'''<span style="color:{color.get(requisition.status)};">{requisition.status}</span>'''

def get_colors_dict(doctype):
    docs = frappe.get_list(doctype, fields=['name', 'color'])
    doc_colors_dict = {}

    for doc in docs:
        doc_colors_dict[doc.name] = doc.color

    return doc_colors_dict


def set_link_to_doc(doc, field, doc_url='',val=None, label=None, color=None):
    style = f'style="color:{color};"' if color else ''
    if val:
        doc[field] = f'''<a href="/app/{doc_url}/{val}" target="_blank" %s>
                        {label or doc[field]}
                    </a>
                ''' % style
    else:
        doc[field] = ''
    return doc[field]

def get_pdf_button(doctype, docname, label='', color=None):
    style = f'style="color:{color};"' if color else ''

    return f'''<a class="btn btn-sm" href="/app/print/{doctype}/{docname}" target="_blank" %s>
                        <svg class="icon  icon-sm" style="">
                            <use class="" href="#icon-printer"></use>
                        </svg>
                    </a>
            ''' % style

def custom_pretty_date(date):
    try:
        if isinstance(date, str):
            if '.' in date:
                date = date.split('.')[0]

            date = datetime.datetime.strptime(date, "%Y-%m-%d %H:%M:%S")

        pretty_date_ = pretty_date(date)

        if pretty_date_:
            pretty_date_ = pretty_date_.split(' ')

            if pretty_date_[1][0:2] == "mo":
                pretty_date_ = f"{pretty_date_[0]}M"

            pretty_date_ = f"{pretty_date_[0]}{pretty_date_[1][0]}"
            return pretty_date_
    except:
        return ""

@frappe.whitelist()
def export_file(**filters):
    columns = get_columns()
    data = get_export_data(filters)
    file_name = 'Requisition_Report.xlsx'
    generate_excel_and_download(columns, data, file_name, height=30)

def get_export_data(filters):
    query_result = get_absolute_data(filters, export=True)
    return query_result

@frappe.whitelist()
def update_field(**kwargs):
    docname = kwargs.get('docname')
    fieldname = kwargs.get('fieldname')
    value = kwargs.get('value')
    leave_application = frappe.get_doc("Merchandising Picture", docname)

    if leave_application.get(fieldname) != value:
        leave_application.update({fieldname:value})
        leave_application.save()

    return True

def change_color(doc, field, color):
    style = f'style="color:{color};"' if color else ''

    doc[field] = f'''<span %s>{doc[field] or "" }</span>''' % style

    return doc[field]

@frappe.whitelist()
def update_delivery_stop_status(trips,value):
    if trips and value:
        try:
            trips = json.loads(trips)
            value = json.loads(value)
        except:
            pass

        for trip in trips:
            trip_split = trip.split("&")
            trip_doc = frappe.get_doc("Delivery Trip",trip_split[0])
            if trip_doc.delivery_stops:
                for stop in trip_doc.delivery_stops:
                    if stop.get("name") == trip_split[1]:
                        stop.status = value.get("status")
            trip_doc.save()